dna/models.py

```diff
@@ -423,67 +423,36 @@ def __init__(
             )
         self.lstm.to(device=device)
 
-        # TODO: use a more standardized/general DAG format
-        self.NULL_INPUTS = ['inputs.0']
+    def forward(self, dags, dag_structure, lstm_start_state):
+        assert len(dag_structure) > 0
+        assert len(dag_structure) == dags.shape[1]
 
-    def forward(self, pipelines, pipeline_structure, lstm_start_state):
-        # TODO: generalize naming so that any DAG could be used
-        seq_len = pipelines.shape[1]
-        assert len(pipeline_structure) == seq_len, 'Pipeline sequence length does not match length of pipeline structure'
-
-        # Keep track of the previous hidden and cell states as we traverse the pipeline
-        prev_lstm_states = []
-
-        lstm_output = None
-
-        # For each batch of primitives coming from the batch of pipelines
-        for i in range(seq_len):
-            # Get the one hot encoded primitives from the batch at index i in the pipeline
-            encoded_primitives = pipelines[:, i, :]
-
-            # Add a dimension in the middle of the shape so it is batch size by sequence length by input size
-            # Note that the sequence length (middle) is 1 because we're doing 1 (batch of) primitive(s) at a time
-            # Also note that the batch size is first in the shape because we selected batch first
-            encoded_primitives = encoded_primitives.unsqueeze(dim=1)
-
-            # Get the list of indices of input primitives coming into the current primitive
-            primitive_inputs = pipeline_structure[i]
-
-            # TODO: generalize so that a node could have the "null" input and inputs from other nodes
-            # If this is the first primitive
-            if primitive_inputs == self.NULL_INPUTS:
-                lstm_input_state = lstm_start_state
-            else:
-                # Otherwise get the mean of the hidden states of the input primitives
-                lstm_input_state = self.get_lstm_input_state(
-                    prev_lstm_states=prev_lstm_states, primitive_inputs=primitive_inputs
-                )
-
-            (lstm_output, lstm_output_state) = self.lstm(encoded_primitives, lstm_input_state)
-
-            prev_lstm_states.append(lstm_output_state)
+        prev_lstm_states = {'inputs.0': lstm_start_state}  # TODO: use something better than 'inputs.0'
+        for i, node_inputs in enumerate(dag_structure):
+            # TODO: use LSTMCell
```

Comment:
What is "LSTMCELL"?

Reply:
https://pytorch.org/docs/stable/nn.html#lstmcell

It is the core component of an LSTM. It is the component that is used at each step in a sequence. It takes as input the hidden state, the cell state, and input feature vector at time step t and outputs  new hidden and cell states as well as the output feature vector.

Reply:
It is essentially an LSTM for a sequence of length 1.

Reply:
Cool